File: typescript/leetcode/19.remove-nth-node-from-end-of-list.ts
/*
 * @lc app=leetcode id=19 lang=typescript
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    let dummy: ListNode = new ListNode();
    dummy.next = head;
    let prev = getNthFromEnd(dummy, n + 1);
    prev.next = prev.next.next;
    return dummy.next;
};

function getNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    let a: ListNode = head;
    for (let i = 0; i < n; i++) {
        a = a.next;
    }
    let b: ListNode = head;
    while (a != null) {
        a = a.next;
        b = b.next;
    }
    return b;
}
// @lc code=end

File: java/leetcode/19.remove-nth-node-from-end-of-list.java
/*
 * @lc app=leetcode id=19 lang=java
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
class Solution {

    public ListNode removeNthFromEnd(ListNode head, int n) {

        ListNode a = head;
        for (int i = 0; i < n; i++) {
            a = a.next;
        }
        if (a == null) {
            return head.next;
        }

        ListNode b = head;
        while (a.next != null) {
            a = a.next;
            b = b.next;
        }
        b.next = b.next.next;

        return head;
    }
}
// @lc code=end
